packages/flutter_tools/lib/src/compile.dart

```diff
@@ -0,0 +1,138 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:flutter_tools/src/artifacts.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/io.dart';
+import 'package:flutter_tools/src/globals.dart';
+import 'package:usage/uuid/uuid.dart';
+
+String _dartExecutable() {
+  final String engineDartSdkPath = artifacts.getArtifactPath(
+    Artifact.engineDartSdkPath
+  );
+  return fs.path.join(engineDartSdkPath, 'bin', 'dart');
+}
+
+Future<String> compile({String sdkRoot, String mainPath}) async {
+  final String frontendServer = artifacts.getArtifactPath(
+    Artifact.frontendServerSnapshotForEngineDartSdk
+  );
+
+  if (!sdkRoot.endsWith('/'))
+    sdkRoot = "$sdkRoot/";
+  final Process server = await Process.start(_dartExecutable(),
+    <String>[frontendServer, '--sdk-root', sdkRoot, mainPath]
+  );
+
+  String boundaryKey;
+  String outputFilename;
+  server.stderr
+    .transform(UTF8.decoder)
+    .listen((String s) { print(s); });
+  server.stdout
+    .transform(UTF8.decoder)
+    .transform(new LineSplitter())
+    .listen((String string) async {
+      const String RESULT_PREFIX = 'result ';
+      if (boundaryKey == null) {
+        if (string.startsWith(RESULT_PREFIX))
+          boundaryKey = string.substring(RESULT_PREFIX.length);
+      } else {
+        if (string.startsWith(boundaryKey)) {
+          if (string.length > boundaryKey.length)
+            outputFilename = string.substring(boundaryKey.length + 1);
+        }
+        else {
+          print('compile debug message: $string');
+        }
+      }
+    });
+  await server.exitCode;
+  return outputFilename;
+}
+
+class ResidentCompiler {
+  ResidentCompiler(this._sdkRoot) {
+    assert(_sdkRoot != null);
+    if (!_sdkRoot.endsWith('/'))
+      _sdkRoot = "$_sdkRoot/";
+  }
+
+  String _sdkRoot;
+  Process _server;
+  Completer<String> _outputFilename;
+  String _boundaryKey;
+
+  Future<String> recompile(List<String> invalidatedFiles) async {
+    if (_server == null) {
+      throw new Future<String>.error("Recompile should be preceded by compile");
```

Comment:
I think returning `Future.error` is cleaner because it seems to accommodate both Future API `future.then(...).catchError(...)` as well as `try { await future; } catch {...}`.
Throwing from the future doesn't seem to work well with `catchError`.
See https://dartpad.dartlang.org/db35be4475a56bd7f4ef57e19ccf2618 for example.